rfcs/Create-object-service.md

```diff
@@ -0,0 +1,53 @@
+# RFC  - Create an object service
+* Updates: [RFC 120](https://github.com/taskcluster/taskcluster-rfcs/blob/master/rfcs/0120-artifact-service.md)
+* Obsoletes: None
+* Comments: [#116](https://github.com/taskcluster/taskcluster-rfcs/pull/116), [#120](https://github.com/taskcluster/taskcluster-rfcs/pull/120)
+* Initially Proposed by: @jhford
+
+# Summary
+Taskcluster tasks often produce artifacts during the course of being run.
+Examples include browser distribution archives, json metadata and log files.
+We currently store artifacts in a single region of S3 and split management of
+artifacts between the Queue and allegedly Cloud Mirror.  
+
+This project is to design an object management api which pulls that functionality 
+out of the Queue and enables support of multiple clouds.
+
+This RFC is to decide on what the objects service is and its interface, not
+implementation details.
+
+# Motivation
+
+* We want to pay less, so we want to be able to store things in different regions and clouds
+* We want the Queue to be easier to understand, develop and maintain
+* We want object management to be easier to understand, develop and maintain
+* We want better object handling (including security, integrity and de-duplication) without 
+increasing complexity of the Queue codebase
+
+# Details
+
+This is to be implemented as a separate service which is to be included into Taskcluster monorepo.
+At the time of writing, this is imagined as a 2-component mini-system: the service itself, which exposes the API,
+and a CLI tool that users can install and use locally.
+
+The service is to be implemented in such a way that it could be as easily as possible switched from Azure to Postgres.
+The relational database for the service should be designed as well (in the form of an entity relationship diagram as part
+of the documentation, if Postgres support will not be deployed by the time of the prototype development).
+
+The service should be compatible with the current implementation (the object storage code in the Queue service), 
+and require as little changes to the clients and workers as possible (preferably, none at all).
```

Comment:
It might be worth following a similar approach to what dustin/hassan are doing with azure support, i.e. breaking out the object-handling portion of the queue into a standalone service without making any changes at all, and then iterating on that new service.

Reply:
> It might be worth following a similar approach to what dustin/hassan are doing with azure support, i.e. breaking out the object-handling portion of the queue into a standalone service without making any changes at all, and then iterating on that new service.

Makes sense to me! I'll start hacking on this as a part of my planning. I'll also time-box writing a little script that puts an object into S3 and gets a hash of the data back. I need to do a bit of hacking because it would help me to plan and to understand how much work this is. I'll time-box the script tho to prevent sliding into a rabbit hole. Separating the functionality first and making sure it works the same way (and defining "the same way") and then changing it makes sense to me as an overall plan.